port/unix/omrsysinfo.c

```diff
@@ -1431,18 +1440,193 @@ retrieveLinuxMemoryStats(struct OMRPortLibrary *portLibrary, struct J9MemoryInfo
+	}
+
+	/* Read value of page cache memory from memory.stat file */
+	memStatFs = getHandleOfCgroupSubsystemFile(portLibrary, OMR_CGROUP_SUBSYSTEM_MEMORY, CGROUP_MEMORY_STAT_FILE);
+	if (NULL == memStatFs) {
+		goto _exit;
+	}
+
+	while (0 == feof(memStatFs)) {
+		char statEntry[MAX_LINE_LENGTH] = {0};
+		char *tmpPtr = NULL;
+
+		if (NULL == fgets((char *)statEntry, MAX_LINE_LENGTH, memStatFs)) {
+			break;
+		}
+		tmpPtr = (char *)statEntry;
+
+		/* Extract "cache" value */
+		if (0 == strncmp(tmpPtr, CGROUP_MEMORY_STAT_CACHE, CGROUP_MEMORY_STAT_CACHE_SZ)) {
+			tmpPtr += CGROUP_MEMORY_STAT_CACHE_SZ;
+			rc = sscanf(tmpPtr, "%" SCNu64, &cgroupMemInfo->cached);
+			if (1 != rc) {
+				Trc_PRT_retrieveLinuxCgroupMemoryStats_invalidValue(CGROUP_MEMORY_STAT_CACHE, CGROUP_MEMORY_STAT_FILE);
+				rc = portLibrary->error_set_last_error_with_message_format(portLibrary, OMRPORT_ERROR_SYSINFO_CGROUP_SUBSYSTEM_FILE_INVALID_VALUE, "invalid value for field %s in file %s", CGROUP_MEMORY_STAT_CACHE, CGROUP_MEMORY_STAT_FILE);
+			}
+			break;
+		}
+	}
+
+_exit:
+	if (NULL != memStatFs) {
+		fclose(memStatFs);
+	}
+
+	return rc;
+}
+
+/**
+ * Function collects memory usage statistics on Linux platforms and returns the same.
+ * This function takes into account cgroup limits if available.
+ *
+ * @param[in] portLibrary The port library.
+ * @param[in] memInfo A pointer to the J9MemoryInfo struct which we populate with memory usage.
+ *
+ * @return 0 on success and -1 on failure.
+ */
+static int32_t
+retrieveLinuxMemoryStats(struct OMRPortLibrary *portLibrary, struct J9MemoryInfo *memInfo)
+{
+	int32_t rc = 0;
+	uint64_t maxVirtual = 0;
+#if !defined(OMRZTPF)
+	struct OMRCgroupMemoryInfo cgroupMemInfo = {0};
+	BOOLEAN isCgroupMemUsageValid = FALSE;
+	BOOLEAN isCgroupMemAndSwapLimitValid = FALSE;
+	uint64_t swapLimit = 0;
+#endif /* !defined(OMRZTPF) */
+	Trc_PRT_retrieveLinuxMemoryStats_Entered();
+
 	rc = portLibrary->sysinfo_get_limit(portLibrary, OMRPORT_RESOURCE_ADDRESS_SPACE | OMRPORT_LIMIT_HARD, &maxVirtual);
 	if ((OMRPORT_LIMIT_UNKNOWN != rc) && (RLIM_INFINITY != maxVirtual)) {
 		memInfo->totalVirtual = maxVirtual;
 	}
-	/* Reset return code to 0, since we are successful in retrieving memory usage. Failure
-	 * conditions would have jumped over to _cleanup section and shall return an appropriately
-	 * set error code.
+
+	rc = retrieveLinuxMemoryStatsFromProcFS(portLibrary, memInfo);
+	if (0 != rc) {
+		goto _exit;
+	}
+
+#if !defined(OMRZTPF)
+	if (OMRPORT_MEMINFO_NOT_AVAILABLE == memInfo->totalPhysical) {
+		goto _exit;
+	}
+
+	/* If cgroup memory subsystem is enabled, retrieve memory limits of the cgroup */
+	if (!portLibrary->sysinfo_cgroup_are_subsystems_enabled(portLibrary, OMR_CGROUP_SUBSYSTEM_MEMORY)) {
+		goto _exit;
+	}
+
+	/* Before retrieving stats from cgroup, back up some host memory stats
+	 * which may be used for calculating available physical memory on the host.
 	 */
-	rc = 0;
+	memInfo->hostAvailPhysical = memInfo->availPhysical;
```

Comment:
Can this code  be part of `retrieveLinuxMemoryStatsFromProcFS` so both `hostXXXX` & `XXXXX` are set at the same time?

Reply:
Right, we need to set it even if there is no cgroup. Fixed that.